File: frontend/lib/features/habits/data/models/requests/habit.dart
import 'dart:collection';

class HabitUpdateRequestModel {
  final Map<String, String> name;
  final Map<String, String> description;
  final String categoryId;
  final bool reviewed;
  final String icon;
  final HashSet<String> unitIds;

  const HabitUpdateRequestModel({
    required this.name,
    required this.description,
    required this.categoryId,
    required this.reviewed,
    required this.icon,
    required this.unitIds,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'category_id': categoryId,
      'reviewed': reviewed,
      'icon': icon,
      'unit_ids': unitIds.toList(),
    };
  }
}

class HabitCreateRequestModel {
  final Map<String, String> name;
  final Map<String, String> description;
  final String categoryId;
  final String icon;
  final HashSet<String> unitIds;

  const HabitCreateRequestModel({
    required this.name,
    required this.description,
    required this.categoryId,
    required this.icon,
    required this.unitIds,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'category_id': categoryId,
      'icon': icon,
      'unit_ids': unitIds.toList(),
    };
  }
}

File: backend/api/src/features/habits/structs/models/habit.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::prelude::FromRow;
use uuid::Uuid;

pub const HABIT_DESCRIPTION_MAX_LENGTH: usize = 2_000;

#[derive(Debug, Deserialize, Serialize, Clone, FromRow)]
pub struct Habit {
    pub id: Uuid,
    pub name: String,
    pub category_id: Uuid,
    pub reviewed: bool,
    pub description: String,
    pub icon: String,
    pub created_at: DateTime<Utc>,
    pub unit_ids: String,
}

#[derive(Serialize, Debug, Deserialize, Clone)]
pub struct HabitData {
    pub id: Uuid,
    pub name: String,
    pub category_id: Uuid,
    pub reviewed: bool,
    pub description: String,
    pub icon: String,
    pub unit_ids: String,
}

impl Habit {
    pub fn to_habit_data(&self) -> HabitData {
        HabitData {
            id: self.id,
            name: self.name.to_owned(),
            category_id: self.category_id,
            reviewed: self.reviewed,
            description: self.description.to_owned(),
            icon: self.icon.to_owned(),
            unit_ids: self.unit_ids.to_owned(),
        }
    }
}
